### PR TITLE
feat(cli): skip payment and upload for existing chunks

### PR DESCRIPTION
**Pull request by maqi**
### Summary generated by Reviewpad on 12 Dec 23 13:13 UTC
This pull request includes the following changes:

1. Added an environment variable "SAFE_DATA_PATH" and set its value to "/home/runner/.local/share/safe".
2. Added a step to start a different client to upload the same file using a different client. This step includes various commands to perform the upload and display the output.
3. Added a step for testing data integrity during nodes churn.
4. Made changes in the `sn_cli` module to handle the verification of chunks and payment for storage.
5. Made changes in the `sn_client` module to handle payment for chunks and storage, and added a function to skip existing chunks during payment.
6. Made changes in the `sn_client` module to handle payment for storage in the `ClientRegister` struct.
7. Made changes in the `sn_client` module to add a list of existing chunks that need to be skipped during payment for storage.
8. Made changes in the `sn_client` module to return a list of skipped chunks during payment for storage.
9. Made changes in the `sn_node` module to handle storage payment proofs cached in the wallet.
10. Updated the test cases in the `storage_payments` module to reflect the changes in payment for storage.

Overall, these changes improve the handling of payment for storage and the verification of chunks in the code.

### CODE REVIEW

File: .github/workflows/memcheck.yml [hunk 1]
```diff
@@ -101,6 +102,21 @@ jobs:
...
+      - name: Start a different client to upload the same file
+        run: |
+          mv $CLIENT_DATA_PATH $NODE_DATA_PATH/client_1
+          cargo run --bin faucet --release -- --log-output-dest=data-dir send 5000000 $(cargo run --bin safe --release -- --log-output-dest=data-dir wallet address | tail -n 1) > initial_balance_from_faucet_1.txt
+          cat initial_balance_from_faucet_1.txt
+          cat initial_balance_from_faucet_1.txt | tail -n 1 > transfer_hex
+          cat transfer_hex
+          cargo run --bin safe --release -- --log-output-dest=data-dir wallet receive --file transfer_hex
+          cargo run --bin safe --release -- --log-output-dest=data-dir files upload "./the-test-data.zip" --show-holders > second_upload.txt
+          cat second_upload.txt
```
Comment (joshuef) [inline .github/workflows/memcheck.yml:116]: can we make this a full test and check the output here has no spend? (or am I missing that?)

File: sn_cli/src/subcommands/files/mod.rs [hunk 2]
```diff
@@ -167,6 +167,10 @@ async fn upload_files(
     if chunk_manager.is_chunks_empty() {
         // make sure we don't have any failed chunks in those
         let chunks = chunk_manager.already_put_chunks(&files_path)?;
+        println!(
+            "Files tried to be uploaded previously, verifying {} chunks",
```
Comment (joshuef) [inline sn_cli/src/subcommands/files/mod.rs]:
```suggestion
            "Files upload attempted previously, verifying {} chunks",
```

File: sn_client/src/file_apis.rs [hunk 3]
```diff
@@ -188,11 +188,11 @@ impl Files {
     pub async fn pay_for_chunks(
         &self,
         chunks: Vec<XorName>,
-    ) -> Result<(NanoTokens, NanoTokens, NanoTokens)> {
+    ) -> Result<((NanoTokens, NanoTokens, NanoTokens), Vec<XorName>)> {
         let mut wallet_client = self.wallet()?;
         info!("Paying for and uploading {:?} chunks", chunks.len());
 
-        let (storage_cost, royalties_fees) =
+        let ((storage_cost, royalties_fees), skipped) =
```
Comment (joshuef) [inline sn_client/src/file_apis.rs]: `skipped_chunks` perhaps for clarity?

**Review by joshuef — APPROVED** [review]

File: .github/workflows/memcheck.yml (hunk 1 above)
Comment (maqi) [reply .github/workflows/memcheck.yml:116]: added an `rg` check to confirm the balance of wallet remain unchanged

**Review by joshuef — APPROVED** [review]

**Pull request merged**